src/main/java/uk/gov/justice/digital/converter/avro/AvroToSparkSchemaConverter.java

```diff
@@ -36,7 +43,36 @@ private Schema toAvroSchema(String avro) {
         // The Avro Schema Parser is stateful and will cache parsed schemas. We choose not to benefit from this since
         // the parser also prevents an existing definition from being updated and will throw a SchemaParseException.
         // This allows us to handle new versions of a schema as and when they are published.
-        return new Schema.Parser().parse(avro);
+        val parsedAvroSchema = new Schema.Parser().parse(avro);
+        // Before returning the schema we must apply our nullable property to any fields declared as nullable.
+        val updatedFields = applyNullablePropertyToFields(parsedAvroSchema);
+
+        return Schema.createRecord(
+                parsedAvroSchema.getName(),
+                parsedAvroSchema.getDoc(),
+                parsedAvroSchema.getNamespace(),
+                parsedAvroSchema.isError(),
+                updatedFields
+        );
+    }
+
+    // Ensure our custom nullable: true|false field is represented in the correct avro form (union type with null) so
+    // that conversion to Spark results in a StructType with the correct nullability properties.
+    private List<Schema.Field> applyNullablePropertyToFields(Schema avro) {
+        return avro.getFields().stream()
+                .map(this::applyNullablePropertyToField)
+                .collect(Collectors.toList());
+    }
+
+    private Schema.Field applyNullablePropertyToField(Schema.Field avroField) {
+
+        val isNullable = Optional.ofNullable(avroField.getObjectProp("nullable"))
+                .map(nullable -> nullable.equals(true))
```

Comment:
Do you need this map?

Reply:
Sadly yes. `getObjectProp` returns `Object` so the map implicitly gives us the boolean we need.